From df5128224084d380408daf8836115cc1a6996f21 Mon Sep 17 00:00:00 2001
From: Madara273 <ravenhoxs@gmail.com>
Date: Thu, 27 Feb 2025 21:21:07 +0000
Subject: [PATCH 07/15] net-tcp_bbr: v2: record app-limited status of
 TLP-repaired flight

When sending a TLP retransmit, record whether the outstanding flight
of data is application limited. This is important for congestion
control modules that want to respond to losses repaired by TLP
retransmits. This is important because the following scenarios convey
very different information:
 (1) a packet loss with a small number of packets in flight;
 (2) a packet loss with the maximum amount of data in flight allowed
     by the CC module;

adapted from -- ref: https://github.com/ferstar/xiaomi_xaga_kernel/commit/8c71dae5648914a6b7054e5ca9bb600f61facefb

Signed-off-by: Madara273 <ravenhoxs@gmail.com>
---
 include/linux/tcp.h   | 3 ++-
 net/ipv4/tcp_output.c | 1 +
 2 files changed, 3 insertions(+), 1 deletion(-)

diff --git a/include/linux/tcp.h b/include/linux/tcp.h
index 6eee6d1ece5ff..4618566f107b9 100644
--- a/include/linux/tcp.h
+++ b/include/linux/tcp.h
@@ -272,7 +272,8 @@ struct tcp_sock {
 	u16	advmss;		/* Advertised MSS			*/
 	u8	compressed_ack;
 	u8	tlp_retrans:1,	/* TLP is a retransmission */
-		unused_1:7;
+	tlp_orig_data_app_limited:1, /* app-limited before TLP rtx? */
+		unused_1:6;
 	u8	fast_ack_mode:2; /* which fast ack mode ? */
 	u32	chrono_start;	/* Start time in jiffies of a TCP chrono */
 	u32	chrono_stat[3];	/* Time in jiffies for chrono_stat stats */
diff --git a/net/ipv4/tcp_output.c b/net/ipv4/tcp_output.c
index 034cf4ed21c85..142168467f426 100644
--- a/net/ipv4/tcp_output.c
+++ b/net/ipv4/tcp_output.c
@@ -2703,6 +2703,7 @@ void tcp_send_loss_probe(struct sock *sk)
 	if (WARN_ON(!skb || !tcp_skb_pcount(skb)))
 		goto rearm_timer;
 
+	tp->tlp_orig_data_app_limited = TCP_SKB_CB(skb)->tx.is_app_limited;
 	if (__tcp_retransmit_skb(sk, skb, 1))
 		goto rearm_timer;
 
-- 
2.47.2

